File: frontend/src/types/training.ts
// Training system types

export type TrainingSlot = 'morning' | 'afternoon' | 'evening'
export type TrainingMode = 'team' | 'groups_2' | 'groups_3'
export type TrainingStatus = 'planned' | 'completed' | 'cancelled'

export interface TrainingItem {
  id: string
  name: string
  category: string
  recommended_group: string
  base_gain: number
  intensity: string
  fitness_delta: number
  fatigue_delta: number
  load_points: number
  attribute_weights: Record<string, number>
  position_fit: Record<string, number>
  is_recovery: boolean
}

export interface TrainingGroup {
  group_id: string
  name: string
  training_item_id: string
  player_ids: string[]
}

export interface TrainingPlanSlot {
  id?: string
  team_id: string
  season_id: string
  season_day: number
  slot: TrainingSlot
  mode: TrainingMode
  training_item_id: string | null
  groups: TrainingGroup[] | null
  status: TrainingStatus
  created_by: string
  training_item?: TrainingItem
}

export interface TrainingTemplate {
  id: string
  name: string
  description: string
}

/** 前端使用的模板详情（含7天schedule） */
export interface TrainingTemplateDetail {
  id: string
  name: string
  description: string
  schedule: string[][] // 7天 x 最多3时段的训练项ID
}

/** 前端本地计划单元格元数据 */
export interface LocalPlanCell {
  training_item_id: string | null
  isAutoSuggested: boolean
  isUserModified: boolean
  isMatchDay: boolean
}

/** 训练计划分组（前端使用） */
export interface PlanGroup {
  group_id: string
  name: string
  player_ids: string[]
  training_item_id: string | null
}

/** 训练计划时段数据（前端使用） */
export interface PlanSlotData {
  mode: TrainingMode
  training_item_id: string | null
  groups: PlanGroup[] | null
  isAutoSuggested: boolean
  isUserModified: boolean
  isMatchDay: boolean
}

export interface TrainingResultItem {
  id: string
  player_id: string
  player_name: string | null
  season_day: number
  slot: TrainingSlot
  training_item_id: string
  training_item_name: string | null
  attribute_gains: Record<string, number>
  fitness_before: number
  fitness_after: number
  fatigue_before: number
  fatigue_after: number
  breakthroughs: Array<{
    attribute: string
    from: number
    to: number
  }>
  efficiency: number
  created_at: string
}

export interface TrainingDailySummary {
  season_day: number
  slot: TrainingSlot
  total_players: number
  total_breakthroughs: number
  breakthrough_players: Array<{
    player_id: string
    player_name: string
    attribute: string
    from: number
    to: number
  }>
}

export interface PlayerFatigueItem {
  player_id: string
  player_name: string
  avatar_url?: string
  fitness: number
  fatigue: number
  stamina_preview: number
  fatigue_band: string
  stamina_multiplier: number
  recommendation: string
  can_high_intensity: boolean
}

export interface TeamFatigueResponse {
  team_id: string
  players: PlayerFatigueItem[]
  avg_fitness: number
  avg_fatigue: number
}

export interface PlayerTrainingProgress {
  player_id: string
  player_name: string
  recent_sessions: number
  total_gains: Record<string, number>
  attribute_status: Record<string, {
    current: number
    cap: number
    progress_pct: number
    status: string
  }>
  growth_curve: Record<string, unknown>
}

export interface AutoGroupResponse {
  mode: TrainingMode
  groups: TrainingGroup[]
}

// Category display helpers
export const TRAINING_CATEGORY_COLORS: Record<string, string> = {
  '战术': 'bg-[#59C7EE]/15 text-[#1F5F43] border-[#59C7EE]/40',
  '技术': 'bg-[#FF6F59]/12 text-[#173126] border-[#FF6F59]/40',
  '恢复': 'bg-[#B9EF3F]/20 text-[#173126] border-[#B9EF3F]/60',
  'tactic': 'bg-[#59C7EE]/15 text-[#1F5F43] border-[#59C7EE]/40',
  'technical': 'bg-[#FF6F59]/12 text-[#173126] border-[#FF6F59]/40',
  'recovery': 'bg-[#B9EF3F]/20 text-[#173126] border-[#B9EF3F]/60',
}

export const TRAINING_CATEGORY_BG: Record<string, { bg: string; border: string; text: string }> = {
  '战术': { bg: 'rgba(89,199,238,0.12)', border: 'rgba(89,199,238,0.35)', text: 'text-[#1F5F43]' },
  '技术': { bg: 'rgba(255,111,89,0.10)', border: 'rgba(255,111,89,0.30)', text: 'text-[#173126]' },
  '恢复': { bg: 'rgba(185,239,63,0.15)', border: 'rgba(185,239,63,0.40)', text: 'text-[#173126]' },
  'tactic': { bg: 'rgba(89,199,238,0.12)', border: 'rgba(89,199,238,0.35)', text: 'text-[#1F5F43]' },
  'technical': { bg: 'rgba(255,111,89,0.10)', border: 'rgba(255,111,89,0.30)', text: 'text-[#173126]' },
  'recovery': { bg: 'rgba(185,239,63,0.15)', border: 'rgba(185,239,63,0.40)', text: 'text-[#173126]' },
}

// ==================== 训练套餐模板（与后端 training_config.py 保持一致）====================

export const TRAINING_TEMPLATES: TrainingTemplateDetail[] = [
  {
    id: 'standard_microcycle',
    name: '标准微周期',
    description: '新手默认，攻防、传控、身体和恢复都有覆盖',
    schedule: [
      ['rondo_4v2', 'first_touch_escape', 'mobility_session'],
      ['box_finish_one_touch', 'delay_and_channel', 'match_review_unit'],
      ['repeat_sprint', 'line_breaking_pass', 'full_rest'],
      ['build_up_2_3', 'central_compactness', 'hydro_recovery'],
      ['dribble_cone_tight', 'standing_tackle_timing', 'role_meeting'],
      ['game_model_8v8', 'corner_near_post', 'full_rest'],
      ['full_rest', 'mobility_session', 'opponent_clip_study'],
    ],
  },
  {
    id: 'finishing_week',
    name: '禁区终结周',
    description: '提升射门、镇定、远射、爆发力，适合进球效率低的球队',
    schedule: [
      ['box_finish_one_touch', 'near_post_finish', 'mobility_session'],
      ['box_finish_under_pressure', 'cutback_finish', 'match_review_unit'],
      ['weak_foot_finish', 'long_shot_window', 'full_rest'],
      ['volley_second_ball', 'penalty_routine', 'hydro_recovery'],
      ['box_finish_one_touch', 'penalty_pressure', 'role_meeting'],
      ['cutback_finish', 'far_post_arrival', 'full_rest'],
      ['full_rest', 'mobility_session', 'opponent_clip_study'],
    ],
  },
  {
    id: 'possession_week',
    name: '控球出球周',
    description: '提升传球、控球、视野、球商，适合中后场控球和出球',
    schedule: [
      ['rondo_4v2', 'first_touch_escape', 'mobility_session'],
      ['third_man_combination', 'switch_play', 'match_review_unit'],
      ['line_breaking_pass', 'back_to_goal_link', 'full_rest'],
      ['build_out_under_press', 'receiving_scanning', 'hydro_recovery'],
      ['wall_pass_timing', 'transition_attack', 'role_meeting'],
      ['game_model_8v8', 'build_up_2_3', 'full_rest'],
      ['full_rest', 'mobility_session', 'opponent_clip_study'],
    ],
  },
  {
    id: 'high_press_week',
    name: '高压反抢周',
    description: '提升体能、抢断、防守意识、球商、爆发力，强度高',
    schedule: [
      ['press_trigger', 'counterpress_after_loss', 'mobility_session'],
      ['cover_shadow_press', 'recovery_run', 'match_review_unit'],
      ['repeat_sprint', 'change_direction', 'full_rest'],
      ['accel_5m', 'transition_defense', 'hydro_recovery'],
      ['standing_tackle_timing', 'delay_and_channel', 'role_meeting'],
      ['game_model_8v8', 'aerobic_blocks', 'full_rest'],
      ['full_rest', 'hydro_recovery', 'opponent_clip_study'],
    ],
  },
  {
    id: 'recovery_week',
    name: '密集赛程恢复周',
    description: '多恢复和低强度认知课，适合连续比赛后调整',
    schedule: [
      ['full_rest', 'hydro_recovery', 'mobility_session'],
      ['recovery_bike', 'match_review_unit', 'opponent_clip_study'],
      ['individual_treatment', 'role_meeting', 'full_rest'],
      ['mobility_session', 'captain_meeting', 'hydro_recovery'],
      ['full_rest', 'opponent_clip_study', 'recovery_bike'],
      ['match_review_unit', 'mobility_session', 'full_rest'],
      ['full_rest', 'hydro_recovery', 'individual_treatment'],
    ],
  },
]

/** 获取模板的默认训练项ID（按天偏移和时段索引） */
export function getTemplateItemId(template: TrainingTemplateDetail, dayOffset: number, slotIndex: number): string | null {
  const daySchedule = template.schedule[dayOffset % template.schedule.length]
  if (!daySchedule) return null
  return daySchedule[slotIndex] ?? null
}


export interface TrainingProgressPoint {
  season_day: number
  value: number
}

export interface TrainingProgressBreakthrough {
  season_day: number
  attribute: string
  before: number
  after: number
}

export interface TrainingProgressSeries {
  player_id: string
  player_name: string
  avatar_url?: string
  values: TrainingProgressPoint[]
  breakthroughs: TrainingProgressBreakthrough[]
}

export interface TrainingProgressResponse {
  metric: string
  metric_label: string
  start_day: number
  end_day: number
  series: TrainingProgressSeries[]
}

File: backend/app/schemas/training.py
"""
Training-related schemas
"""
from typing import Optional, List, Dict
from decimal import Decimal
from datetime import datetime
from pydantic import Field
from enum import Enum

from app.schemas.base import BaseSchema


class TrainingSlot(str, Enum):
    MORNING = "morning"
    AFTERNOON = "afternoon"
    EVENING = "evening"


class TrainingMode(str, Enum):
    TEAM = "team"
    GROUPS_2 = "groups_2"
    GROUPS_3 = "groups_3"


class TrainingItemSchema(BaseSchema):
    """训练内容 schema"""
    id: str
    name: str
    category: str
    recommended_group: str
    base_gain: float
    intensity: str
    fitness_delta: int
    fatigue_delta: int
    load_points: int
    attribute_weights: Dict[str, float] = Field(default_factory=dict)
    position_fit: Dict[str, float] = Field(default_factory=dict)
    is_recovery: bool = False


class TrainingGroupSchema(BaseSchema):
    """分组配置 schema"""
    group_id: str
    name: str
    training_item_id: str
    player_ids: List[str]


class TrainingPlanSlotSchema(BaseSchema):
    """训练计划时段 schema"""
    id: Optional[str] = None
    team_id: str
    season_id: str
    season_day: int
    slot: TrainingSlot
    mode: TrainingMode = TrainingMode.TEAM
    training_item_id: Optional[str] = None
    groups: Optional[List[TrainingGroupSchema]] = None
    status: str = "planned"
    created_by: str = "player"
    training_item: Optional[TrainingItemSchema] = None


class TrainingPlanSaveRequest(BaseSchema):
    """保存训练计划请求"""
    season_id: str
    items: List[dict] = Field(..., description="训练计划项列表")


class TemplateApplyRequest(BaseSchema):
    """套用套餐请求"""
    season_id: str
    start_day: int


class TrainingResultSchema(BaseSchema):
    """训练结果 schema"""
    id: str
    player_id: str
    player_name: Optional[str] = None
    season_day: int
    slot: TrainingSlot
    training_item_id: str
    training_item_name: Optional[str] = None
    attribute_gains: Dict[str, float] = Field(default_factory=dict)
    fitness_before: int
    fitness_after: int
    fatigue_before: int
    fatigue_after: int
    breakthroughs: List[dict] = Field(default_factory=list)
    efficiency: Decimal
    created_at: datetime


class TrainingDailySummarySchema(BaseSchema):
    """每日训练总结"""
    season_day: int
    slot: TrainingSlot
    total_players: int
    total_breakthroughs: int
    breakthrough_players: List[dict] = Field(default_factory=list)


class PlayerFatigueSchema(BaseSchema):
    """球员疲劳状态 schema"""
    player_id: str
    player_name: str
    fitness: int
    fatigue: int
    stamina_preview: float
    fatigue_band: str
    stamina_multiplier: float
    recommendation: str
    can_high_intensity: bool


class PlayerTrainingProgressSchema(BaseSchema):
    """球员训练进度 schema"""
    player_id: str
    player_name: str
    recent_sessions: int
    total_gains: Dict[str, float] = Field(default_factory=dict)
    attribute_status: Dict[str, dict] = Field(default_factory=dict)
    growth_curve: dict = Field(default_factory=dict)


class AutoGroupResponse(BaseSchema):
    """自动分组响应"""
    mode: TrainingMode
    groups: List[TrainingGroupSchema]


class TrainingTemplateSchema(BaseSchema):
    """训练套餐 schema"""
    id: str
    name: str
    description: str


class PlayerFatigueBatchResponse(BaseSchema):
    """全队疲劳状态响应"""
    team_id: str
    players: List[PlayerFatigueSchema]
    avg_fitness: float
    avg_fatigue: float



class TrainingProgressPoint(BaseSchema):
    """折线图单点"""
    season_day: int
    value: float


class TrainingProgressBreakthrough(BaseSchema):
    """整数突破标记"""
    season_day: int
    attribute: str
    before: int
    after: int


class TrainingProgressSeries(BaseSchema):
    """单个球员在某项指标上的折线序列"""
    player_id: str
    player_name: str
    avatar_url: Optional[str] = None
    values: List[TrainingProgressPoint] = Field(default_factory=list)
    breakthroughs: List[TrainingProgressBreakthrough] = Field(default_factory=list)


class TrainingProgressResponse(BaseSchema):
    """训练成长曲线响应"""
    metric: str
    metric_label: str
    start_day: int
    end_day: int
    series: List[TrainingProgressSeries] = Field(default_factory=list)
